cypher-shell/src/main/java/org/neo4j/shell/parser/ShellStatementParser.java

```diff
@@ -0,0 +1,188 @@
+package org.neo4j.shell.parser;
+
+import javax.annotation.Nonnull;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+import java.util.regex.Pattern;
+
+/**
+ * A cypher aware parser which can detect shell commands (:prefixed) or cypher.
+ */
+public class ShellStatementParser implements StatementParser {
+
+    private static final Pattern shellCmdPattern = Pattern.compile("^\\s*:.+\\s*$");
+    private static final char SEMICOLON = ';';
+    private static final char BACKSLASH = '\\';
+    private static final String LINE_COMMENT_START = "//";
+    private static final String LINE_COMMENT_END = "\n";
+    private static final String BLOCK_COMMENT_START = "/*";
+    private static final String BLOCK_COMMENT_END = "*/";
+    private static final char BACKTICK = '`';
+    private static final char DOUBLE_QUOTE = '"';
+    private static final char SINGLE_QUOTE = '\'';
+    private Optional<String> awaitedRightDelimiter;
+    private StringBuilder statement;
+    private ArrayList<String> parsedStatements;
+
+
+    public ShellStatementParser() {
+        parsedStatements = new ArrayList<>();
+        statement = new StringBuilder();
+        awaitedRightDelimiter = Optional.empty();
+    }
+
+    @Override
+    public void parseMoreText(@Nonnull String line) {
+        // See if it could possibly be a shell command, only valid if not in a current statement
+        if (!statementStarted() && shellCmdPattern.matcher(line).find()) {
+            parsedStatements.add(line);
+            return;
+        }
+
+        // We will guess it is cypher then
+        boolean skipNext = false;
+        char prev, current = (char) 0;
+        for (char c : line.toCharArray()) {
+            // append current
+            statement.append(c);
+            // last char shuffling
+            prev = current;
+            current = c;
+
+            if (skipNext) {
+                // This char is escaped so gets no special treatment
+                skipNext = false;
+                continue;
+            }
+
+            // Handle comments
```

Comment:
Some of the comments could be extracted to methods Ex: `handleComments` it makes the method read smaller.